File: src/test/java/ar/com/flow/bankaccount/SavingsAccountTest.java
package ar.com.flow.bankaccount;

import ar.com.flow.bankaccount.balance.Balance;
import ar.com.flow.bankaccount.transfer.SameAccountException;
import ar.com.flow.money.InsufficientFundsException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static ar.com.flow.bankaccount.BankAccountAssert.assertThat;
import static ar.com.flow.bankaccount.OptionalTransactionRecordAssert.assertThat;
import static ar.com.flow.bankaccount.TestObjects.*;
import static ar.com.flow.money.TestObjects.*;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class SavingsAccountTest {
    private BankAccount franciscosAccount;
    private BankAccount mabelsAccount;

    @BeforeEach
    void setUp() {
        franciscosAccount = createSavingsAccountFor(francisco, dollars100);
        mabelsAccount = createSavingsAccountFor(mabel, dollars100);
    }

    @Test
    void createdAccountHasBasicInformation() {
        var account = SavingsAccount.builder().owner(francisco).currency("ARS").build();

        assertEquals(francisco, account.getOwner());
        assertEquals("ARS", account.getCurrency());
        assertEquals(Balance.zero("ARS"), account.getBalance());
    }

    @Test
    void depositIncrementsFunds() {
        franciscosAccount.deposit(dollars10);

        assertThat(franciscosAccount).increasedFunds(dollars10);
    }

    @Test
    void withdrawalDecrementsFunds() {
        franciscosAccount.withdraw(dollars10);

        assertThat(franciscosAccount).decreasedFunds(dollars10);
    }

    @Test
    void canNotWithdrawAmountGreaterThanAvailableFunds() {
        assertThrows(InsufficientFundsException.class, () -> franciscosAccount.withdraw(dollars200));

        assertThat(franciscosAccount).keepsInitialBalance();
    }

    @Test
    void withdrawalAddsTransactionRecordToStatement() {
        franciscosAccount.withdraw(dollars10);

        assertThat(franciscosAccount.getStatement().last())
                .isPresent()
                .isWithdrawal()
                .hasNegativeBalance(dollars10);
    }

    @Test
    void transferAmountLessThanAvailableFunds() {
        franciscosAccount.transfer(dollars10, mabelsAccount);

        assertThat(franciscosAccount).decreasedFunds(dollars10);

        assertThat(mabelsAccount).increasedFunds(dollars10);

        assertThat(franciscosAccount.getStatement().last())
                .isPresent()
                .isDebit()
                .isTransfer()
                .hasNegativeBalance(dollars10)
                .hasCreditAccount(mabelsAccount);

        assertThat(mabelsAccount.getStatement().last())
                .isPresent()
                .isCredit()
                .isTransfer()
                .hasPositiveBalance(dollars10);
    }

    @Test
    void canNotTransferAmountGreaterThanAvailableFunds() {
        assertThrows(InsufficientFundsException.class, () ->
                franciscosAccount.transfer(dollars200, mabelsAccount));

        assertThat(franciscosAccount).keepsInitialBalance();
        assertThat(mabelsAccount).keepsInitialBalance();
    }

    @Test
    void canNotTransferToSelf() {
        assertThrows(SameAccountException.class, () ->
                franciscosAccount.transfer(dollars10, franciscosAccount));

        assertThat(franciscosAccount).keepsInitialBalance();
    }
}

File: src/test/kotlin/ar/com/flow/bankaccount/SavingsAccountTest.kt
package ar.com.flow.bankaccount

import ar.com.flow.bankaccount.balance.Balance.Companion.zero
import ar.com.flow.bankaccount.transfer.SameAccountException
import ar.com.flow.money.InsufficientFundsException
import ar.com.flow.money.TestObjects.dollars10
import ar.com.flow.money.TestObjects.dollars200
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class SavingsAccountTest {
    private lateinit var franciscosAccount: BankAccount
    private lateinit var mabelsAccount: BankAccount

    @BeforeEach
    fun setUp() {
        franciscosAccount =
            TestObjects.createSavingsAccountFor(TestObjects.francisco, ar.com.flow.money.TestObjects.dollars100)
        mabelsAccount = TestObjects.createSavingsAccountFor(TestObjects.mabel, ar.com.flow.money.TestObjects.dollars100)
    }

    @Test
    fun createdAccountHasBasicInformation() {
        val account = SavingsAccount(TestObjects.francisco, "ARS")
        assertEquals(TestObjects.francisco, account.owner)
        assertEquals("ARS", account.currency)
        assertEquals(zero("ARS"), account.balance)
    }

    @Test
    fun depositIncrementsFunds() {
        franciscosAccount.deposit(dollars10)
        BankAccountAssert.assertThat(franciscosAccount)
            .increasedFunds(dollars10)
    }

    @Test
    fun withdrawalDecrementsFunds() {
        franciscosAccount.withdraw(dollars10)
        BankAccountAssert.assertThat(franciscosAccount)
            .decreasedFunds(dollars10)
    }

    @Test
    fun canNotWithdrawAmountGreaterThanAvailableFunds() {
        Assertions.assertThrows(InsufficientFundsException::class.java) { franciscosAccount.withdraw(dollars200) }
        BankAccountAssert.assertThat(franciscosAccount).keepsInitialBalance()
    }

    @Test
    fun withdrawalAddsTransactionRecordToStatement() {
        franciscosAccount.withdraw(dollars10)
        OptionalReceiptAssert.assertThat(franciscosAccount.statement.last())
            .isPresent()
            .isWithdrawal()
            .hasNegativeBalance(dollars10)
    }

    @Test
    fun transferAmountLessThanAvailableFunds() {
        franciscosAccount.transfer(dollars10, mabelsAccount)
        BankAccountAssert.assertThat(franciscosAccount)
            .decreasedFunds(dollars10)
        BankAccountAssert.assertThat(mabelsAccount).increasedFunds(dollars10)
        OptionalReceiptAssert.assertThat(franciscosAccount.statement.last())
            .isPresent()
            .isDebit()
            .isTransfer()
            .hasNegativeBalance(dollars10)
            .hasCreditAccount(mabelsAccount)
        OptionalReceiptAssert.assertThat(mabelsAccount.statement.last())
            .isPresent()
            .isCredit()
            .isTransfer()
            .hasPositiveBalance(dollars10)
    }

    @Test
    fun canNotTransferAmountGreaterThanAvailableFunds() {
        Assertions.assertThrows(InsufficientFundsException::class.java) {
            franciscosAccount.transfer(dollars200, mabelsAccount)
        }
        BankAccountAssert.assertThat(franciscosAccount).keepsInitialBalance()
        BankAccountAssert.assertThat(mabelsAccount).keepsInitialBalance()
    }

    @Test
    fun canNotTransferToSelf() {
        Assertions.assertThrows(SameAccountException::class.java) {
            franciscosAccount.transfer(dollars10, franciscosAccount)
        }
        BankAccountAssert.assertThat(franciscosAccount).keepsInitialBalance()
    }
}